text/0000-gradual-struct-init.md

```diff
@@ -0,0 +1,739 @@
+`struct` with `#[non_exhaustive]` in one crate, then a downstream crate won't
+be able to use the gradual syntax to make a value of the type.
+
+### No `Drop` types
+
+Consider a type which implements `Drop` by printing something (17):
+
+```rust
+struct SpecialDrop {
+    alpha: u8,
+    beta: u16,
+    gamm: u32,
+}
+
+impl Drop for SpecialDrop {
+    fn drop(&mut self) {
+        println!("Dropping in a special way!");
+    }
+}
+```
+
+Let's then use the gradual initialization mechanism in this RFC (18):
+
+```rust
+fn foo() {
+    let sd: SpecialDrop;
+
+    sd.alpha = 1;
+    sd.beta = 2;
+
+    some_action_that_can_panic();
+
+    sd.gamma = 3;
+}
+```
+
+If you don't have the definition of `SpecialDrop` in your near vicinity,
+it can become difficult to know whether the destructor, and thus the side effect
+in `SpecialDrop::drop` will run or when it will run. In (18), to know if and
+when the destructor will run, you would need to know that `SpecialDrop` has the
+fields `alpha`, `beta`, and `gamma` and that `gamma` hasn't been initialized yet.
+
+Thus, while it would be sound to allow it, if a type implements `Drop`,
+the compiler will reject will not allow gradual initialization.
+
+### Tuples and tuple structs also work
+
+Hitherto, we have only gradually initialized `struct`s with named fields.
+However, this RFC makes no distinction between those and other forms of
+heterogeneous product types. In other words, **_you can also initialize
+tuple structs and normal tuples with the gradualist syntax._**
+That is, you can write (19):
+
+```rust
+struct Point(f32, f32);
+
+let pt: Point;
+
+pt.0 = 42.24;
+pt.1 = 13.37;
+
+drop(pt);
+```
+
+as well as (20):
+
+```rust
+fn consume_pair((x, y): (u8, u8)) { ... }
+
+let pt;
+pt.0 = 42;
+pt.1 = 24;
+
+consume_pair(pt);
+```
+
+### Gradual initialization in fields
+
+What we have said so far does not just apply to local bindings.
+You can use the gradualist mechanism in fields as well.
+For example, you may use the gradual notation as a sort of builder notation (21):
+
+```rust
+struct Config {
+    window: WindowConfig,
+    runtime: RuntimeConfig,
+}
+
+struct WindowConfig {
+    height: usize,
+    width: usize,
+}
+
+struct RuntimeConfig {
+    threads: usize,
+    max_memory: usize,
+}
+
+fn make_config() -> Config
+    let cfg;
+
+    cfg.window.width = 1920;
+    cfg.window.height = 1080;
+    // cfg.window is definitely initialized now.
+
+    cfg.runtime.threads = 8;
+    cfg.runtime.max_memory = 1024;
+    // cfg.runtime is definitely initialized now.
+    // therefore cfg is also.
+
+    cfg
+}
+```
+
+The usual same rules with respect to eventual definite initialization,
+privacy, uninhabited types, prohibitions on `Drop` types, etc. apply
+here as well.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+There are no changes to the grammar.
+
+## Semantics
```

Comment:
@pnkfelix & @nikomatsakis: also please review this section so that it makes sense to you and reflects the guide section / or if not let's make it do so...

Reply:
cc @RalfJ so I "get a place in the queue of stuff" Ralf has to review. =P